tracing-syslog/src/syslog.rs

```diff
@@ -0,0 +1,220 @@
+use std::{borrow::Cow, ffi::CStr};
+use tracing_core::{
+    field::{Field, Visit},
+    Collect, Event, Level,
+};
+use tracing_subscriber::{
+    registry::{LookupSpan, Registry},
+    subscribe::{CollectExt, Context, Subscribe},
+};
+
+/// `syslog` options.
+#[derive(Copy, Clone, Debug, Default)]
+pub struct Options(libc::c_int);
+
+impl Options {
+    /// Log the pid with each message.
+    pub const LOG_PID: Self = Self(libc::LOG_PID);
+    /// Log on the console if errors in sending.
+    pub const LOG_CONS: Self = Self(libc::LOG_CONS);
+    /// Delay open until first syslog() (default).
+    pub const LOG_ODELAY: Self = Self(libc::LOG_ODELAY);
+    /// Don't delay open.
+    pub const LOG_NDELAY: Self = Self(libc::LOG_NDELAY);
+    /// Don't wait for console forks: DEPRECATED.
+    pub const LOG_NOWAIT: Self = Self(libc::LOG_NOWAIT);
+    /// Log to stderr as well.
+    pub const LOG_PERROR: Self = Self(libc::LOG_PERROR);
+}
+
+impl std::ops::BitOr for Options {
+    type Output = Self;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        Self(self.0 | rhs.0)
+    }
+}
+
+/// `syslog` facility.
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone, Debug)]
+#[repr(i32)]
+pub enum Facility {
+    /// Kernel messages (these can't be generated from user processes).
+    LOG_KERN = libc::LOG_KERN,
+    /// Generic user-level messages.
+    LOG_USER = libc::LOG_USER,
+    /// Mail subsystem.
+    LOG_MAIL = libc::LOG_MAIL,
+    /// System daemons without separate facility value.
+    LOG_DAEMON = libc::LOG_DAEMON,
+    /// Security/authorization messages.
+    LOG_AUTH = libc::LOG_AUTH,
+    /// Messages generated internally by `syslogd(8)`.
+    LOG_SYSLOG = libc::LOG_SYSLOG,
+    /// Line printer subsystem.
+    LOG_LPR = libc::LOG_LPR,
+    /// USENET news subsystem.
+    LOG_NEWS = libc::LOG_NEWS,
+    /// UUCP subsystem.
+    LOG_UUCP = libc::LOG_UUCP,
+    /// Clock daemon (`cron` and `at`).
+    LOG_CRON = libc::LOG_CRON,
+    /// Security/authorization messages (private).
+    LOG_AUTHPRIV = libc::LOG_AUTHPRIV,
+    /// FTP daemon.
+    LOG_FTP = libc::LOG_FTP,
+    /// Reserved for local use.
+    LOG_LOCAL0 = libc::LOG_LOCAL0,
+    /// Reserved for local use.
+    LOG_LOCAL1 = libc::LOG_LOCAL1,
+    /// Reserved for local use.
+    LOG_LOCAL2 = libc::LOG_LOCAL2,
+    /// Reserved for local use.
+    LOG_LOCAL3 = libc::LOG_LOCAL3,
+    /// Reserved for local use.
+    LOG_LOCAL4 = libc::LOG_LOCAL4,
+    /// Reserved for local use.
+    LOG_LOCAL5 = libc::LOG_LOCAL5,
+    /// Reserved for local use.
+    LOG_LOCAL6 = libc::LOG_LOCAL6,
+    /// Reserved for local use.
+    LOG_LOCAL7 = libc::LOG_LOCAL7,
+}
+
+impl Default for Facility {
+    fn default() -> Self {
+        Self::LOG_USER
+    }
+}
+
+/// `syslog` severity.
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone)]
+#[repr(i32)]
+pub enum Severity {
```

Comment:
In the past, we've typically [used consts instead of enums for levels](https://github.com/tokio-rs/tracing/blob/master/tracing-core/src/metadata.rs#L263-L286) for two reasons:

1. `#[non_exhaustive]` enums were not stable.
2. The ability for us to change the representation of a level without introducing a breaking change.

That being said, the first reason no longer holds, but the second one does. However, I'm not sure how to achieve what you're doing doing here with our typical approach—Eliza can probably give more information. That being said, can you add a `#[non_exhaustive]` to this enum if you _are_ going to stick with an enum?

Reply:
See my comment below about just making it private. But regardless, I think it's pretty safe to say that the set of `syslog` severities will never change, so what would be the point of making it `#[non_exhaustive]`?

Reply:
Made it private